kubernetes/resource_kubernetes_namespace.go

```diff
@@ -87,6 +113,23 @@ func resourceKubernetesNamespaceUpdate(d *schema.ResourceData, meta interface{})
 	ctx := context.TODO()
 
 	ops := patchMetadata("metadata.0.", "/metadata/", d)
+
+	if d.HasChange("spec.0.finalizers") {
+		r := d.Get("spec.0.finalizers")
+
+		finalizers := make([]v1.FinalizerName, 0)
+		for _, f := range r.([]interface{}) {
+			finalizer := v1.FinalizerName(f.(string))
+			finalizers = append(finalizers, finalizer)
+		}
+
+		log.Printf("haha %#v", r)
+		ops = append(ops, &ReplaceOperation{
+			Path:  "/spec/finalizers",
+			Value: finalizers,
+		})
+	}
```

Comment:
Hmm, this looks correct offhand. But I see what you mean, it's not actually working. Here are logs from applying this with `TF_LOG=trace`.

https://gist.githubusercontent.com/dak1n1/915bc7e682a14e5c8a838fd8b3dc40a6/raw/db9d6c16048ee0a040801892f9fca6a34451e71f/gistfile1.txt

You can see the request was submitted successfully, but the object remains unchanged (just one finalizer listed instead of two).

This seems to be enforced by the Kubernetes API itself, as you can see by this manual test:

```
[dakini@dax 20201005]$ kube edit namespace test -v7
I1005 12:10:58.003446   38864 loader.go:375] Config loaded from file:  /home/dakini/.kube/config
I1005 12:10:58.004348   38864 cert_rotation.go:137] Starting client certificate rotation controller
I1005 12:10:58.007521   38864 round_trippers.go:420] GET https://192.168.39.49:8443/api/v1/namespaces/test
I1005 12:10:58.007532   38864 round_trippers.go:427] Request Headers:
I1005 12:10:58.007537   38864 round_trippers.go:431]     Accept: application/json
I1005 12:10:58.007540   38864 round_trippers.go:431]     User-Agent: kubectl/v1.19.0 (linux/amd64) kubernetes/e199641
I1005 12:10:58.016199   38864 round_trippers.go:446] Response Status: 200 OK in 8 milliseconds
I1005 12:10:58.016799   38864 editor.go:128] Opening file with editor [vi /tmp/kubectl-edit-as522.yaml]
I1005 12:11:15.031662   38864 editoptions.go:298] User edited:
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2020-10-05T19:04:17Z"
  name: test
  resourceVersion: "1097"
  selfLink: /api/v1/namespaces/test
  uid: 6aa1777e-f0ef-464c-886a-81aa2af0a622
spec:
  finalizers:
  - kubernetes
  - foregroundDeletion
status:
  phase: Active
I1005 12:11:15.032998   38864 round_trippers.go:420] GET https://192.168.39.49:8443/openapi/v2?timeout=32s
I1005 12:11:15.033043   38864 round_trippers.go:427] Request Headers:
I1005 12:11:15.033072   38864 round_trippers.go:431]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I1005 12:11:15.033097   38864 round_trippers.go:431]     User-Agent: kubectl/v1.19.0 (linux/amd64) kubernetes/e199641
I1005 12:11:15.062323   38864 round_trippers.go:446] Response Status: 200 OK in 29 milliseconds
I1005 12:11:15.187899   38864 round_trippers.go:420] PATCH https://192.168.39.49:8443/api/v1/namespaces/test?fieldManager=kubectl-edit
I1005 12:11:15.187912   38864 round_trippers.go:427] Request Headers:
I1005 12:11:15.187916   38864 round_trippers.go:431]     Content-Type: application/strategic-merge-patch+json
I1005 12:11:15.187920   38864 round_trippers.go:431]     User-Agent: kubectl/v1.19.0 (linux/amd64) kubernetes/e199641
I1005 12:11:15.187923   38864 round_trippers.go:431]     Accept: application/json
I1005 12:11:15.191600   38864 round_trippers.go:446] Response Status: 200 OK in 3 milliseconds
namespace/test edited
```
```
[dakini@dax 20201005]$ kube get namespace test -o json |jq
{
  "apiVersion": "v1",
  "kind": "Namespace",
  "metadata": {
    "creationTimestamp": "2020-10-05T19:04:17Z",
    "managedFields": [
      {
        "apiVersion": "v1",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:status": {
            "f:phase": {}
          }
        },
        "manager": "HashiCorp",
        "operation": "Update",
        "time": "2020-10-05T19:04:43Z"
      },
      {
        "apiVersion": "v1",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:spec": {
            "f:finalizers": {}
          }
        },
        "manager": "kubectl-edit",
        "operation": "Update",
        "time": "2020-10-05T19:11:15Z"
      }
    ],
    "name": "test",
    "resourceVersion": "1371",
    "selfLink": "/api/v1/namespaces/test",
    "uid": "6aa1777e-f0ef-464c-886a-81aa2af0a622"
  },
  "spec": {
    "finalizers": [
      "kubernetes"
    ]
  },
  "status": {
    "phase": "Active"
  }
}
```

Based on this, I think this PR won't work after all. It probably makes more sense to pursue a metadata-level finalizer instead. https://github.com/hashicorp/terraform-provider-kubernetes/issues/855#issuecomment-676732693